test/sub_doc_functions.js

```diff
@@ -0,0 +1,175 @@
+    for(var model in thinky.models) {
+      modelNameSet[model] = true;
+    }
+    modelNames = Object.keys(modelNameSet);
+    thinky._clean();
+    done();
+  });
+}
+
+
+
+describe('Sub Document Functions', function(){
+
+  afterEach(cleanTables);
+
+  var basicModelOneSpec = {
+    id: String,
+    name: String,
+    email: type.string(),
+    address: {
+      street: String,
+      geoCode: function() {
+        return this.street;
+      }
+    },
+    topLevelFunc: function() {
+      return "topLevel";
+    }
+  }
+  var basicModelOne = null;
+
+  it('can create a basic model', function(done){
+    var testModelOne = thinky.createModel("TestModelOne", basicModelOneSpec);
+    assert.equal(type.isObject(testModelOne._schema._schema.address), true);
+    assert.equal(typeof testModelOne._schema._schema.address._methods, 'object');
+    assert.equal(Object.keys(testModelOne._schema._schema.address._methods).length, 1);
+    assert.equal(testModelOne._schema._schema.address._methods.geoCode, basicModelOneSpec.address.geoCode);
+    basicModelOne = testModelOne;
+    done();
+  });
+
+  var basicModelTwoSpec = {
+    id: String,
+    name: String,
+    email: type.string(),
+    addresses: [{
+      street: String,
+      geoCode: function() {
+        return this.street;
+      }
+    }]
+  }
+  var basicModelTwo = null;
+
+  it('can create a basic model with object array', function(done){
+    var testModelTwo = thinky.createModel("TestModelTwo", basicModelTwoSpec);
+    assert.equal(type.isArray(testModelTwo._schema._schema.addresses), true);
+    assert.equal(typeof testModelTwo._schema._schema.addresses._schema._methods, 'object');
+    assert.equal(Object.keys(testModelTwo._schema._schema.addresses._schema._methods).length, 1);
+    assert.equal(testModelTwo._schema._schema.addresses._schema._methods.geoCode, basicModelTwoSpec.addresses[0].geoCode);
+    basicModelTwo = testModelTwo;
+    done();
+  });
+
+  it('can instantiate basic objects correctly', function(done){
+    var one =  new basicModelOne({
+      id: 'db15e340-1ebf-4c24-871b-c383330cef7f',
+      name: 'Tester',
+      email: 'tester@test.com',
+      address: {
+        street: "123 Fake Street"
+      }
+    });
+    assert.equal(typeof one.address, 'object');
+    assert.equal(one.topLevelFunc(), 'topLevel');
+    assert.equal(one.address.geoCode(), "123 Fake Street");
+    var two =  new basicModelTwo({
+      id: '7777b2c1-26ac-45e0-a0f2-f11e613265b9',
+      name: 'Tester2',
+      email: 'tester2@test.com',
+      addresses: [{
+        street: "456 Fake Street"
+      }]
+    });
+    assert.equal(Array.isArray(two.addresses), true);
+    assert.equal(two.addresses.length, 1);
+    assert.equal(two.addresses[0].geoCode(), "456 Fake Street");
+    done();
+  });
+
+  var basicModelThreeSpec = {
+    id: String,
+    name: String,
+    email: type.string(),
+    address: {
+      street: String,
+      geoCode: function() {
+        return this.street;
```

Comment:
The difference is that (from my understanding of the doco), `define` can only add functions to the top level model object, not sub-docs/sub-objects.

In the case of `basicModelThreeSpec `, `address` is a sub-object of the model that owns the function `geocode`, rather than the function having to be defined onto the model itself.

The changes also apply to arrays of sub-objects by allowing each object to have functions added to them based on the schema definition